src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs

```diff
@@ -184,6 +194,27 @@ public static bool IsDisposeImplementation([NotNullWhen(returnValue: true)] this
                 method.IsImplementationOfInterfaceMethod(null, iDisposable, "Dispose");
         }
 
+        /// <summary>
+        /// Checks if the given method implements "IAsyncDisposable.Dispose" or overrides an implementation of "IAsyncDisposable.Dispose".
+        /// </summary>
+        public static bool IsAsyncDisposeImplementation([NotNullWhen(returnValue: true)] this IMethodSymbol? method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? iAsyncDisposable, INamedTypeSymbol? valueTaskType)
+        {
+            if (method == null)
+            {
+                return false;
+            }
+
+            if (method.IsOverride)
+            {
+                return method.OverriddenMethod.IsAsyncDisposeImplementation(iAsyncDisposable, valueTaskType);
+            }
+
+            // Identify the implementor of IAsyncDisposable.Dispose in the given method's containing type and check
+            // if it is the given method.
+            return method.ReturnType.Equals(valueTaskType) && method.Parameters.Length == 0 &&
```

Comment:
Are we sure that `ReturnType` is never null?

```suggestion
            return method.ReturnType != null && method.ReturnType.Equals(valueTaskType) && method.Parameters.Length == 0 &&
```

If you apply the change you can also add the `[NotNullWhen(returnValue: true)]` on the `valueTaskType` parameter.

Reply:
I can't think of any case that can make ReturnType to be null. Do you have any case in mind?

Reply:
Some error, also maybe a void type.

Reply:
@Evangelink if that's the case, it would be a problem with the existing implementation too right?

Reply:
Yeah they are checking for `ReturnsVoid` :)